src/Utilities/functions.jl

```diff
@@ -135,45 +165,80 @@ If `x` (resp. `y`, `z`) is `VariableIndex(1)` (resp. 2, 3).
 The canonical representation of `ScalarAffineFunction([y, x, z, x, z], [2, 1, 3, -2, -3], 5)` is `ScalarAffineFunction([x, y], [-1, 2], 5)`.
 
 """
-function canonical(f::SAF{T}) where T
-    sorted_terms = sort(f.terms, by = t -> t.variable_index.value)
-    terms = MOI.ScalarAffineTerm{T}[]
-    for t in sorted_terms
-        if !isempty(terms) && t.variable_index == last(terms).variable_index
-            terms[end] = MOI.ScalarAffineTerm(terms[end].coefficient + t.coefficient, t.variable_index)
-        elseif !iszero(t.coefficient)
-            if !isempty(terms) && iszero(last(terms).coefficient)
-                terms[end] = t
-            else
-                push!(terms, t)
-            end
-        end
+function canonical(f::MOI.ScalarAffineFunction)
+    if iscanonical(f)
+        f
+    else
+        canonical!(MOI.ScalarAffineFunction(copy(f.terms), f.constant))
     end
-    if !isempty(terms) && iszero(last(terms).coefficient)
-        pop!(terms)
+end
+
+function canonical(f::MOI.VectorAffineFunction)
+    if iscanonical(f)
+        f
+    else
+        canonical!(MOI.VectorAffineFunction(copy(f.terms), f.constants))
```

Comment:
Should probably copy `f.constants` too? This part was there before, but might as well change it.

Reply:
Yeah, I was on the fence about that. The constants are never modified during canonicalization, but I guess they might be changed by something downstream. 

I could just implement: 

```
Base.copy(f::ScalarAffineFunction) = <copy terms and constant>
Base.copy(f::VectorAffineFunction) = <copy terms and constants>
```

and then re-write `canonical(f) = canonicalize!(copy(f))` which would work for both